sign/mayo/mode1/internal/mayo.go

```diff
@@ -0,0 +1,728 @@
+	var tab [len(Tail) * 4]byte
+	for i := 0; i < len(Tail); i++ {
+		tab[4*i] = mul(Tail[i], 1)
+		tab[4*i+1] = mul(Tail[i], 2)
+		tab[4*i+2] = mul(Tail[i], 4)
+		tab[4*i+3] = mul(Tail[i], 8)
+	}
+
+	const lsb = 0x1111111111111111
+
+	for c := 0; c < OKpadded; c += 16 {
+		for r := M; r < M+(K+1)*K/2; r++ {
+			pos := (r/16)*OKpadded + c + (r % 16)
+			t0 := a[pos] & lsb
+			t1 := (a[pos] >> 1) & lsb
+			t2 := (a[pos] >> 2) & lsb
+			t3 := (a[pos] >> 3) & lsb
+			for t := 0; t < len(Tail); t++ {
+				a[((r+t-M)/16)*OKpadded+c+((r+t)%16)] ^= t0*uint64(tab[4*t+0]) ^ t1*uint64(tab[4*t+1]) ^ t2*uint64(tab[4*t+2]) ^ t3*uint64(tab[4*t+3])
+			}
+		}
+	}
+
+	// transform the temporary matrix into the desired form of A matrix
+	KO1 := K*O + 1
+	for r := 0; r < M; r += 16 {
+		for c := 0; c < KO1-1; c += 16 {
+			for i := 0; i < 16; i++ {
+				src := viewUint64SliceAsBytes(a[r/16*OKpadded+c+i:])
+				offset := KO1*(r+i) + c
+				decode(src, _a[offset:offset+min(16, KO1-1-c)])
+			}
+		}
+	}
+}
+
+func transpose16x16Nibbles(m []uint64) {
+	const evenNibbles = 0x0f0f0f0f0f0f0f0f
+	const evenBytes = 0x00ff00ff00ff00ff
+	const even2Bytes = 0x0000ffff0000ffff
+	const evenHalf = 0x00000000ffffffff
+
+	for i := 0; i < 16; i += 2 {
+		t := ((m[i] >> 4) ^ m[i+1]) & evenNibbles
+		m[i] ^= t << 4
+		m[i+1] ^= t
+	}
+
+	for i := 0; i < 16; i += 4 {
+		t0 := ((m[i] >> 8) ^ m[i+2]) & evenBytes
+		t1 := ((m[i+1] >> 8) ^ m[i+3]) & evenBytes
+		m[i] ^= (t0 << 8)
+		m[i+1] ^= (t1 << 8)
+		m[i+2] ^= t0
+		m[i+3] ^= t1
+	}
+
+	for i := 0; i < 4; i++ {
+		t0 := ((m[i] >> 16) ^ m[i+4]) & even2Bytes
+		t1 := ((m[i+8] >> 16) ^ m[i+12]) & even2Bytes
+
+		m[i] ^= t0 << 16
+		m[i+8] ^= t1 << 16
+		m[i+4] ^= t0
+		m[i+12] ^= t1
+	}
+
+	for i := 0; i < 8; i++ {
+		t := ((m[i] >> 32) ^ m[i+8]) & evenHalf
+		m[i] ^= t << 32
+		m[i+8] ^= t
+	}
+}
+
+func Verify(msg []byte, sig []byte, epk *ExpandedPublicKey) bool {
+	if len(sig) != SignatureSize {
+		panic("sig must be of length SignatureSize")
+	}
+
+	var digest [DigestSize]byte
+
+	h := sha3.NewShake256()
+	_, _ = h.Write(msg[:])
+	_, _ = h.Read(digest[:])
+
+	h.Reset()
+	_, _ = h.Write(digest[:])
+	_, _ = h.Write(sig[SignatureSize-SaltSize : SignatureSize])
+
+	var tenc [M / 2]byte
+	_, _ = h.Read(tenc[:])
+
+	var t [M]byte
+	decode(tenc[:], t[:])
+
+	var s [K * N]byte
+	decode(sig[:(K*N+1)/2], s[:])
+
+	P1 := viewBytesAsUint64Slice(epk.p1[:])
+	P2 := viewBytesAsUint64Slice(epk.p2[:])
+	P3 := viewBytesAsUint64Slice(epk.p3[:])
+
+	// Note: the variable time approach is overall about 30% faster
+	// compute P * S^t = [ P1  P2 ] * [S1] = [P1*S1 + P2*S2]
+	//                   [  0  P3 ]   [S2]   [        P3*S2]
+	var pst [M * N * K / 16]uint64
+	// mulAddMMatXMatTrans(pst[:], P1, s[:], V, V, K, N, true)
+	// mulAddMMatXMatTrans(pst[:], P2, s[V:], V, O, K, N, false)
+	// mulAddMMatXMatTrans(pst[M*V*K/16:], P3, s[V:], O, O, K, N, true)
+	variableTime(pst[:], P1, P2, P3, s[:])
```

Comment:
It's not clear at all how this part matches up with what is written in the spec. You need to give an overview how the computation matches up.